numpy/testing/_private/pytesttester.py

```diff
@@ -0,0 +1,329 @@
+"""
+Nose test running.
+
+This module implements ``test()`` and ``bench()`` functions for NumPy modules.
+
+"""
+from __future__ import division, absolute_import, print_function
+
+import os
+import sys
+import warnings
+
+import numpy as np
+from numpy.compat import basestring
+
+from .utils import suppress_warnings
+
+
+__all__ = ['PytestTester', '_numpy_tester']
+
+
+def _get_package_name(filepath):
+    """
+    Given a path where a package is installed, determine its name.
+
+    Parameters
+    ----------
+    filepath : str
+        Path to a file. If the determination fails, "numpy" is returned.
+
+    """
+    pkg_name = []
+    while 'site-packages' in filepath or 'dist-packages' in filepath:
+        filepath, name = os.path.split(filepath)
+        if name in ('site-packages', 'dist-packages'):
+            break
+        pkg_name.append(name)
```

Comment:
I don't think this is reliable in the presence of `.pth` files.

Reply:
The main problem with doing it the scipy way is tracking the develop vs release switch. However, I think that can be internalized into the test runner.

Reply:
I think one reason develop vs release wasn't internalized before was downstream packages were using it for testing. I think we should stop the practice of supporting downstrean package when we switch to pytest because it has caused back compatibility problems. We should be able to change numpy testing without worrying about downstream.

@njsmith @rgommers Thoughts?

Reply:
Agreed that the new pytest based `test` command should be based on a function/class that's not publicly exposed; it's easy for other packages to provide their own (see SciPy's `_testutils.py`).

Would indeed like to keep develop vs release behavior, that is useful.

Reply:
Although, the whole filtering business is kind of tricky to get right.... let me think some more

Reply:
My conclusion: I think we will need those warning filters in SciPy and other projects, however not having them there now is not super critical, and they can easily be copied over. In the end we need to be free here to change this test runner for NumPy's needs without having to worry about backwards compat, so I agree with Chuck that its best to keep this class private.